cpp/tests/io/json/json_test.cpp

```diff
@@ -3253,6 +3255,144 @@ TEST_F(JsonReaderTest, JsonNestedDtypeFilterWithOrder)
+
+    cudf::io::schema_element dtype_schema{
+      data_type{cudf::type_id::STRUCT},
+      {
+        {"a", {data_type{cudf::type_id::LIST}, {{"element", {dtype<int64_t>()}}}}},
+        {"c2",
+         {data_type{cudf::type_id::LIST},
+          {{"element",
+            {data_type{cudf::type_id::STRUCT},
+             {
+               {"d", {data_type{cudf::type_id::STRING}}},
+             },
+             {{"d"}}}}}}},
+      },
+      {{"a", "c2"}}};
+    in_options.set_dtypes(dtype_schema);
+    cudf::io::table_with_metadata result = cudf::io::read_json(in_options);
+    // Make sure we have column "a":[int64_t]
+    ASSERT_EQ(result.tbl->num_columns(), 2);
+    ASSERT_EQ(result.metadata.schema_info.size(), 2);
+    EXPECT_EQ(result.metadata.schema_info[0].name, "a");
+    ASSERT_EQ(result.metadata.schema_info[0].children.size(), 2);
+    EXPECT_EQ(result.metadata.schema_info[0].children[0].name, "offsets");
+    EXPECT_EQ(result.metadata.schema_info[0].children[1].name, "element");
+    // Make sure we have all null list "c2": [{"d": ""}]
+    EXPECT_EQ(result.metadata.schema_info[1].name, "c2");
+    ASSERT_EQ(result.metadata.schema_info[1].children.size(), 2);
+    EXPECT_EQ(result.metadata.schema_info[1].children[0].name, "offsets");
+    EXPECT_EQ(result.metadata.schema_info[1].children[1].name, "element");
+    ASSERT_EQ(result.metadata.schema_info[1].children[1].children.size(), 1);
+    EXPECT_EQ(result.metadata.schema_info[1].children[1].children[0].name, "d");
+
+    auto const expected0 = [&] {
+      auto const valids = std::vector<bool>{1, 0};
+      auto [null_mask, null_count] =
+        cudf::test::detail::make_null_mask(valids.begin(), valids.end());
+      return cudf::make_lists_column(2,
+                                     size_type_wrapper{0, 1, 1}.release(),
+                                     int64_wrapper{1}.release(),
+                                     null_count,
+                                     std::move(null_mask));
+    }();
+
+    auto const expected1 = [&] {
+      auto const get_structs = [] {
+        auto child = cudf::test::strings_column_wrapper{};
+        return cudf::test::structs_column_wrapper{{child}};
+      };
+      auto const valids = std::vector<bool>{0, 0};
+      auto [null_mask, null_count] =
+        cudf::test::detail::make_null_mask(valids.begin(), valids.end());
+      return cudf::make_lists_column(2,
+                                     size_type_wrapper{0, 0, 0}.release(),
+                                     get_structs().release(),
+                                     null_count,
+                                     std::move(null_mask));
+    }();
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected0, result.tbl->get_column(0).view());
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected1, result.tbl->get_column(1).view());
+  }
+}
+
+TEST_F(JsonReaderTest, NullifyMixedList)
+{
+  using namespace cudf::test::iterators;
+  // test list
+  std::string json_stringl = R"(
```

Comment:
In this example, if we have an input such as `{"c2": [{"b": [1], "c": 123}]}`, would the expected output be `[{null, null}]`?

Reply:
Spark CPU shows this:
```
scala> val df = Seq("""{"c2": [{"b": [1], "c": 123}]}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: string, c: string>>>')").show(false)
+----------------+
|from_json(value)|
+----------------+
|{[{[1], 123}]}  |
+----------------+
```
So the output should be `[{[1], 123}]`. And this PR outputs that result :+1:

Reply:
Note that the first value is `[1]` is a string, not an array of string/number:
```
scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: string, c: string>>>')").printSchema
root
 |-- from_json(value): struct (nullable = true)
 |    |-- c2: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- b: string (nullable = true)
 |    |    |    |-- c: string (nullable = true)
```
So this is due to read mixed type as string.

Reply:
For mismatched nested schema, the second value is still not null:
```

scala> val df = Seq("""{"c2": [{"b": [1], "c": 123}]}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> df.show(false)
+------------------------------+
|value                         |
+------------------------------+
|{"c2": [{"b": [1], "c": 123}]}|
+------------------------------+


scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: struct<d:int>, c: string>>>')").show(false)
+----------------+
|from_json(value)|
+----------------+
|{[{null, 123}]} |
+----------------+
```

Reply:
Even with mixed_type_as_string=false, specifying a schema as string, forces all its rows to be returned as string (even if it has list/struct/string/fixed_width_types). This came from one of requirements from spark.